File: vectors/caip_edge_v1/engines/runner.go
// Copyright 2026 AlgoVoi (chopmob@gmail.com). Apache-2.0.
// caip_edge_v1 Go runner. Correct = \A..\z (RE2 end-of-text). Naive = ^..$ search.
// Go RE2 $ matches end of text only (no trailing-newline exception), so Go does NOT share
// the anchor trap: the naive over-accept count is expected to be 0.
package main

import (
	"bufio"
	"encoding/hex"
	"fmt"
	"os"
	"regexp"
	"strings"
)

const chain = `[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}`

var body = map[string]string{
	"caip2":  chain,
	"caip10": chain + `:[-.%a-zA-Z0-9]{1,128}`,
	"caip19": chain + `/[-a-z0-9]{3,8}:[-.%a-zA-Z0-9]{1,128}(/[-.%a-zA-Z0-9]{1,78})?`,
}

func main() {
	ok := map[string]*regexp.Regexp{}
	naive := map[string]*regexp.Regexp{}
	for k, v := range body {
		ok[k] = regexp.MustCompile(`\A` + v + `\z`)
		naive[k] = regexp.MustCompile(`^` + v + `$`)
	}
	f, err := os.Open("corpus.tsv")
	if err != nil {
		fmt.Println("cannot open corpus.tsv:", err)
		os.Exit(2)
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	n, pass, trap := 0, 0, 0
	for sc.Scan() {
		p := strings.SplitN(sc.Text(), "\t", 3)
		if len(p) < 3 {
			continue
		}
		exp, kind, h := p[0], p[1], p[2]
		b, _ := hex.DecodeString(h)
		s := string(b)
		want := exp == "accept"
		if ok[kind].MatchString(s) == want {
			pass++
		}
		n++
		if exp == "reject" && naive[kind].MatchString(s) {
			trap++
		}
	}
	fmt.Printf("Go(RE2)  correct %d/%d | naive ^..$ over-accepts %d reject-vectors\n", pass, n, trap)
	if pass != n {
		os.Exit(1)
	}
}

File: vectors/caip_edge_v1/engines/Runner.java
// Copyright 2026 AlgoVoi (chopmob@gmail.com). Apache-2.0.
// caip_edge_v1 Java runner. Correct = Pattern.matches() (whole-input anchor). Naive =
// Pattern.find() with ^..$: Java's $ (no MULTILINE) matches before a trailing line
// terminator, so find()+^..$ SHARES the anchor trap (over-accepts trailing-newline vectors),
// exactly like Python. matches() does not, because it requires the entire input consumed.
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HexFormat;
import java.util.regex.Pattern;

public class Runner {
    static final String CHAIN = "[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}";

    static String body(String k) {
        switch (k) {
            case "caip2":  return CHAIN;
            case "caip10": return CHAIN + ":[-.%a-zA-Z0-9]{1,128}";
            default:       return CHAIN + "/[-a-z0-9]{3,8}:[-.%a-zA-Z0-9]{1,128}(/[-.%a-zA-Z0-9]{1,78})?";
        }
    }

    public static void main(String[] a) throws Exception {
        var hf = HexFormat.of();
        int n = 0, pass = 0, trap = 0;
        for (String ln : Files.readAllLines(Path.of("corpus.tsv"), StandardCharsets.UTF_8)) {
            String[] p = ln.split("\t", 3);
            if (p.length < 3) continue;
            String exp = p[0], kind = p[1];
            String s = new String(hf.parseHex(p[2]), StandardCharsets.UTF_8);
            boolean want = exp.equals("accept");
            if (Pattern.matches(body(kind), s) == want) pass++;
            n++;
            if (exp.equals("reject") && Pattern.compile("^" + body(kind) + "$").matcher(s).find()) trap++;
        }
        System.out.printf("Java     correct %d/%d | naive find()+^..$ over-accepts %d reject-vectors%n", pass, n, trap);
        System.exit(pass == n ? 0 : 1);
    }
}
